### PR TITLE
Improve types for `ThrottlingOctokitOptions`

### PR DESCRIPTION
**Pull request by oscard0m**
# Description
To improve typing for `octokitOptions` for `plugin-throttling`

# Context
It can be interesting for:
- deprecating properties/methods 
- auto-complete

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    Bottleneck?: any;
+    id?: string;
+    timeout?: number;
+    connection?: any;
+    onAbuseLimit: LimitHandler;
+    onRateLimit: LimitHandler;
+  };
+}
+
+export type OctokitThrottlingOptions = OctokitOptions | ThrottlingOptions;
```
Comment (oscard0m) [inline src/index.d.ts:17]: A loooong time without touching TS. If I don't do this `Union` the compiler does not allow me to use `ThrottlingOptions` in a place where `OctokitOptions` is expected.

Doing this union we are not making `options.throttle` mandatory which is not good.

I still have to play around, maybe with generics `OctokitOptions<T>`?

File: src/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { OctokitOptions } from "@octokit/core/dist-types/types.d";
```
Comment (wolfy1339) [inline src/index.d.ts:1]: This file should be `types.ts` to remain consistent with other the other repos

File: src/index.d.ts (hunk 1 above)
Comment (wolfy1339) [reply src/index.d.ts:17]: If I understand the code properly, it would be sufficient to use `ThrottlingOptions` directly instead of `OctokitThrottlingOptions`, and then we should rename it to `OctokitThrottlingOptions`

**Review by wolfy1339 — COMMENTED** [review]
Looks good.

Left some comments

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+type LimitHandler = (retryAfter: number, options: any, octokit: any) => void;
```
Comment (wolfy1339) [inline src/index.d.ts:3]:
```suggestion
type LimitHandler = (retryAfter: number, options: ThrottlingOptions, octokit: any) => void;
```

File: src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+export interface ThrottlingOptions extends OctokitOptions {
```
Comment (wolfy1339) [inline src/index.d.ts:5]:
```suggestion
export interface OctokitThrottlingOptions extends OctokitOptions {
```

File: src/index.d.ts (hunk 1 above)
Comment (wolfy1339) [inline src/index.d.ts:17]:
```suggestion

```

File: src/index.d.ts (hunk 3 above)
Comment (wolfy1339) [inline src/index.d.ts:3]:
```suggestion
import { Octokit } from '@octokit/core

type LimitHandler = (retryAfter: number, options: ThrottlingOptions, octokit: Octokit) => void;
```

File: src/index.d.ts (hunk 4 above)
Comment (oscard0m) [reply src/index.d.ts:5]: This is what I tried initially but in `src/test/octokit.ts` I get the following TS Error:

image

**Comment by wolfy1339** [comment]
We can get better types for Bottleneck. They just aren't published to npm for some reason.

We'll have to vendor them in the repo.

https://github.com/SGrondin/bottleneck/blob/master/light.d.ts

As for the options type, I'm not entirely sure.
The only idea, I have is to check if the `throttle` property is present and that might satisfy TS

**Comment by oscard0m** [comment]
> We can get better types for Bottleneck. They just aren't published to npm for some reason.
> 
> We'll have to vendor them in the repo.
> 
> https://github.com/SGrondin/bottleneck/blob/master/light.d.ts
> 
> As for the options type, I'm not entirely sure. The only idea, I have is to check if the `throttle` property is present and that might satisfy TS

I had to create a [TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBA8gxsA9gawJbBmYrEDsDOUAvFAN4BQUVUAhgK7AAWAysAE43AQDmIA-AC5auEAG5K1ek0HCxEqnXwQ2AQW4RcwGfnapc3cdShg2EAG6oIAd3zbd+gNoBdQ9QBGNJQFU2AGztsegbyUL6I3DIURkYAJhBudNxCABQAthD4+DTqQjqB+gCUxAB8UHS4yLiIVriu0VB6AGaIKemZ2RC59txFRKXlldW1IUZWNGy4rRlZOVB5Qb39FVU1ddHKbIhsU+2z84UlZctDawC+a6YAjnQZWkI0IlAA9E-wSGjAACrgGQB0AEoQa63QHAnSYbB4fBrbDpABaeAgASCawciCwOEmc26TiEAxWuHI53I5FAkFgCBQ6GIsmeTygSDmqFSYF8qEaICgEAAHjQWb4ICSydAADLM9AACQeMQFbBpyQkpnYIBUjS42yguDoqTcygANBJ0ZCCEJPoxNsBgGz9BDMfgDVREJSPkI3lTgORFlAzIhUDFxOQ9OrGjQ4NAzRarUFbVCyBImJGBUIotQNDQ3AKYjI3IhEAKHmsAEKIS0C3AQODIGQPORGP3I-Qw5kQRAMGRanXKNZwPDlhCY6siNZ4FQJJRi1LoIQTyXS2XD3D-TgQGfAafi4BS3AyrsSYmnIU-KARktWiAwNgAcQ0ylQcAAPJ8udyuNvCCfS9GMbGAD4U97oDGBClEQEhPjyr4xO+5qnta3BAYQfDHjBn42t+BBQK6zqAeh0Iko05T9ngUBwKYy4iumEC+I+z6QdBiZfsa+DFAqjq4aaKFRmhTGekIPp+uIBG4ERuAkWRXAUbq1HgS+GhQbAk7APebofAhepQAARAmsEQBpxQsYa7H-u6CG8d6vr+uQQkiWJEDkZR0m0XJ9GwYxdpQH+Kk4UxBlsUxppmfxMRxlQ2lWJp5TMqyEDpJoEAxBp4gHuQLwAKK8vy0AAIwCOQpF2RJDnJKQpwFKIdJQEadpYQBGC4VAgCg5CSaUZdFUAAEy5fl9lScVDKcUmZC0EIWVQKcY1lRVVVQjVJn1YAMuTNU86V8m1ADMXXiSuRWkP1iadENI5jiuG4pF6JXqXgS4SadUDJOd42PZNLyVUZH5cfB9VNSl9IrZlUAACy5dZxqVbgo6KCdinJEqbAqmqyhdPkPRkAe3WFb1u3aaWB27UdkOrpdi7kbd92HCVY0TeVL3TSaxmqQt5BAA) to explain the problem: 

Basically the `[option: string]: any` inside `type OctokitOptions` is preventing TypeScript to get the branch for `type ThrottlingOptions` when the shape of `throttling` is wrong.

Any ideas on how to bypass this? Would this imply to rethink `type OctokitOptions`? @wolfy1339 (@G-Rath, master of TS, I invoke you 🌋 )

**Comment by G-Rath** [comment]
If I understand what you're wanting correctly (which is that when `plugin-throttling` is being used, `GithubOptions` should have an additional property `throttle` for options for this plugin), I think you should be using [Declaration Merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html).

Roughly: 

```
interface OctokitOptions {
    authStrategy?: any;
    auth?: any;
    userAgent?: string;
    previews?: string[];
    baseUrl?: string;
    log?: {
        debug: (message: string) => unknown;
        info: (message: string) => unknown;
        warn: (message: string) => unknown;
        error: (message: string) => unknown;
    };
    request?: any //OctokitTypes.RequestRequestOptions;
    timeZone?: string;
    [option: string]: unknown
};

interface OctokitOptions {
  throttle?: ThrottlingOptions['throttle']
}

interface ThrottlingOptions {
  throttle: {
    enabled?: boolean;
    Bottleneck?: any;
    id?: string;
    timeout?: number;
    connection?: any;
    onAbuseLimit: LimitHandler;
    onRateLimit: LimitHandler;
  };
}
```

See [playground](https://www.typescriptlang.org/play?ssl=42&ssc=6&pln=42&pc=23#code/JYOwLgpgTgZghgYwgAgPILAewNbDKgBzGExAGdkBvAKGTuTgFcwALAZTCjkgHMBPAPwAuBiD4BuWvSathoiVLqMy0AII8I4OWU6gek+sgJQIAN2AQA7mW26QPANoBdA-QBGcFQFUoAG1tQeq50vpg8cjSGhgAmEG6MPCIAFAC2EGRkcBoiOoH2AJTIALwAfMiMINggmJYgwVGgMJjJaRlZEDl2PIWl5ZXVtfWGlnBQIC3pmdnIuXo9ZRVVNXWKUdBQmFATbdOzBcUL-cv1AL71JgCOjOlgcnBiyAD0j+hYuGAAKnwE6QB0AEoQK43QHAnSEYikMj1YhpABapAgASCqwcmCIJHGMy6ThEiwGIGoZ2o1DA3xQrxweGK8iej2QWBmwBSBF8wBgfGQEAAHnAWb4ICSyT9kAAZZl4AAS92iAqgNKSUhMnD4qhgkC2yBAjBSbmgABopOjIeQRB8WBswGA2fYIZiyIa6JgMFSwCJKe9qPNkKZMMBopJqKANfAkGgXe87VCqFJWJbrUizRbMFabTwo+QHAByOMphNZpxEknB6ChlDm+NpjMUSIM5OpjoxwyaOBuAXRORuTCYAX3eoAITzApAEAQ2DuYnq-uR9hhzIgmGYcm1uug9QQpBHGExE4UhlIqniKnFKTwIhPUplcvqpH+3AgF7dYolYGlIFla6kxJOQvJyArQ4QKgUAAOKaNAwAIAAPB8XLcpA74UABqZ6NWyAAD7hm8eDVmURRSLBPIIdESH1taqEYtGAj-mRVaUeQyDuhGOH0dCJIwBU26kMgCAmPeoqthAvgwXBxGkZWFEmmQJSKk6rFJhJtqsV6Ii+v6kgcSAXEgDxfGQAJerCYR8GaCRaCnmAUEeixUn6sgABEuYNvZJQyUa8lYa61YqT6foBtQmnabpED8YJRmiaZ4lDpJ9oYZ5kasW5clSWaPlqdETZ1hslgORUzKshAaTgBA0T2ZIP7UM8ACivL8igACMQjULxIX6WFSSUCc+TiHSyDGvaTHYfgrHIIAoOQktVtUFcgABMTUtaFhkdVlgEiJQDAiPVyAnNt3W9f1UKDV5I2ADLkE2PDVfLTQAzPNekPu161OQma19SAh7KA+L7JN6nV2be-HfcgSS-TtYN7c8fUech5FKVJY3ncgl11cgAAsTWBSab0fceL5JMqUCquq0CdHk3RUD+C1tUtT20Y260HkeX0Wf9IB3vpQMgwcFPbbtPWQwdprxTZsVnUAA)

This requires that the type be an `interface` (and in general, why it's generally better for libraries to use interfaces rather than `type` as it allows downstream consumers to enhance the types), so it'll need a change in `@octokit/core`.

Once that change has been made, the final code will be something like this:

```
declare module '@octokit/core' {
  interface OctokitOptions {
    throttle?: ThrottlingOptions;
  }
}

interface ThrottlingOptions {
  enabled?: boolean;
  Bottleneck?: any;
  id?: string;
  timeout?: number;
  connection?: any;
  onAbuseLimit: LimitHandler;
  onRateLimit: LimitHandler;
}
```

(module pathing can be a bit funny, so that might need adjusting but it's easier to do when the interface actually exists in the package - in the meantime [this](https://github.com/jest-community/eslint-plugin-jest/blob/main/src/index.ts#L11-L16) is a good example of this in action)

**Review by wolfy1339 — CHANGES_REQUESTED** [review]

File: src/types.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { OctokitOptions } from "@octokit/core/dist-types/types.d";
+import { Octokit } from "@octokit/core";
+
+declare module "@octokit/core" {
+  interface OctokitOptions {
+    throttle?: ThrottlingOptions;
```
[the remark this answers is not in the document]
Comment (oscard0m) [reply src/types.ts]: Types are getting more complex than expected, it will get me some time to sort them out. I'll keep you posted next week

**Comment by oscard0m** [comment]
> Types are getting more complex than expected, it will get me some time to sort them out. I'll keep you posted next week

Pushed a second iteration of the types. Still having problems with  https://github.com/octokit/plugin-throttling.js/blob/054cefa828fe6bf4db3c8962a2f140ecacebee37/test/octokit.ts#L37

I get the following TS error:

```
Argument of type 'typeof throttling' is not assignable to parameter of type 'OctokitPlugin'.
  Types of parameters 'options' and 'options' are incompatible.
    Type 'OctokitOptions' is not assignable to type '{ throttle: ThrottlingOptionsBase & SecondaryLimitHandler; }'.
      Types of property 'throttle' are incompatible.
        Type 'ThrottlingOptions' is not assignable to type 'ThrottlingOptionsBase & SecondaryLimitHandler'.
          Type 'ThrottlingOptionsBase & AbuseLimitHandler' is not assignable to type 'ThrottlingOptionsBase & SecondaryLimitHandler'.
            Property 'onSecondaryRateLimit' is missing in type 'ThrottlingOptionsBase & AbuseLimitHandler' but required in type 'SecondaryLimitHandler'.ts(2345)
[types.ts(24, 3): ]()'onSecondaryRateLimit' is declared here.
```

Any clue on what could be the issue? @wolfy1339 @G-Rath 

> 💡 [The change in `@octokit/core.js`](https://github.com/octokit/core.js/pull/450) to migrate type `OctokitOptions` to an `interface` [failed 💣](https://github.com/octokit/core.js/runs/5412497631?check_suite_focus=true) , until [the issue](https://github.com/octokit/core.js/issues/451) is fixed, you will need to replace it manually in your local `node_modules` to run the tests.

**Review by wolfy1339 — DISMISSED** [review]

File: src/index.ts [hunk 6]
```diff
@@ -45,21 +48,27 @@ const createGroups = function (Bottleneck, common) {
...
 };
 
-export function throttling(octokit: Octokit, octokitOptions = {}) {
+export function throttling(
+  octokit: Octokit,
+  options: { throttle: ThrottlingOptionsBase & SecondaryLimitHandler }
+): {};
+export function throttling(
+  octokit: Octokit,
+  options: { throttle: ThrottlingOptionsBase & AbuseLimitHandler }
```
[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts:57]: Those are [TS method overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) to make the typing work, do you suggest to remove them?

I tried to do it without that and I was not getting the expected typing: 
image


See [this TS playground ](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgeQMYwga2PAvnAMyghDgCIABCdLHAelWgFMyBuAKFEljgCEIYMADZMAdk1SZCxUmQBGA4WImY27TqJhMoBAIaomKGthjIwMYBFEBnROzgO4AbQjnLogFxxrMKMFEA5gC6XrqiAJ4cjnC6AK4wABYAyr66WgHhAPyhEVGOcYnZMbn2jrHW2gCCAWIwRT5+gXAAPnCxogAmTAT+TB15DmBQTABuwEwA7tb1vv4BTkEtbZ3dvf2lDnK6FQCqUEIzjQFL7V094uvRQhABRQgb0XBdcrEBXgAUIEzW1ro1Xg05gBKOAAXgAfMtMKIIBNRANHv4CBAPl8fn8mADZoEQRCoTC4QjohNdFBPHBPt9fv9vNiArjIe1obD4Q9otpiFBUVSMVijgz8SyEfhWqdVhcEcMAI6xb51HKRB4WL4ALSsTEOcxOK3OfQ4uHU-i0On0hjQGBMZgsVls9wciWIghERQAKglHcI5lb3NZ9eoYOEwIYADKgHAACTCHREUDBFNKw184UqBGNRVEsRAcm0ABpSq5rTYihA5AArCQwPMOagWnBFc20GDsAUjCDAdbsANBuCVF4VUMgCNRmNxu1wOgAKgnGwncAoXSGEjSfXITioxnoYCEr38AFoHYohHNFgADKy98pMAc4d5Ak9wYC2BfDVDLjrFd8TYBCIRwbNwYYQAgEYV38GJCHiWJhgApgRG2QwIAIOAT3XWsYDoLcd1Efd3UPOYTxzOAtyYeC2gqOBEkMM9RCSCQrA6UlwgAJWXa8YHvBJh20B8bC0XQOjIGc6HzUQL37MMYC8NjI06GN9Q4TtA0MWjGE6RjpK42NQTsasaLotSoGY1iJKkiSZOjbR5PUJgAA9uHgLtDDdD0j0Cb0bV4UjtLHMRdDkEQOiKBQIDg1kHH4J1lEkIowkVBx201QI8mVJgIHidNM2zKA8lU8R0HcGKSgcQdRFATMVPoxiWK0JimCTFM0y8DMs0shM6sMhrtE8ioADVdG3DUmsy1qHAmPwtCKCKlDyzAADoAHFiFiMA8gqUlUASSbDyiubFrSlbShhCwelfQtpi8KaRBmhaloO3TqqvEy4A02TWtwBTHLgZy8LctwbTjb6nS9P6bG6wwADJSneMTHsHGBzJHVoKoM8IXosqAgQUgh2nyqw4E416oHeGtGwVQiCx9LwG0tEHrBBMdEyg1kDXYOg6AoGBrF3Wyg3QbmoE5dgCfR94EFwQixwPSKvDHe7jLhj4BTFqtHF8-y+i8XxZRV3SYbYxWwUhZWHisZGGMMh79YpJWDQcA1cCBdRhZjUXxbsNmOa5nmK35wX7Vw6WdMcKxLae29DcQcWHjVgLNagbXSntx2hc012JdKKWlBlk3RFDhXrYj432VEPzY4o+OmB1uBzz7WGcANvEi7gJOndT5Wg8zkRs+iEP5frgvG6j4vS418uE57vTVPNoytCt8PB5VpOgA)

Any better suggestion or something I'm missing?

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts:57]: I didn't notice that they were overloads.

I removed them, and it works for me. Are you sure you patched your `node_modules`?

[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts:57]: > I didn't notice that they were overloads.
> 
> I removed them, and it works for me. Are you sure you patched your `node_modules`?

This is the file I modified:

image

Is the right one, right?

Also, in the TS Playground I attached before, I'm "hardcoding" the double interface declaration but it does not work. Is not a good example to check?

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts:57]: That is the file that should be modified, mine looks the same.

This is what the `throttling` function looks like for me:
image

When I hover the `octokitOptions.throttle` option, I get the proper types, as seen in the image
image

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts:57]: This is the complete diff of my changes
```diff
diff --git a/src/index.ts b/src/index.ts
index 99e4d91..ce1b561 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -48,14 +48,6 @@ const createGroups = function (Bottleneck, common) {
   });
 };
 
-export function throttling(
-  octokit: Octokit,
-  options: { throttle: ThrottlingOptionsBase & SecondaryLimitHandler }
-): {};
-export function throttling(
-  octokit: Octokit,
-  options: { throttle: ThrottlingOptionsBase & AbuseLimitHandler }
-): {};
 export function throttling(octokit: Octokit, octokitOptions: OctokitOptions) {
   const {
     enabled = true,
@@ -113,7 +105,7 @@ export function throttling(octokit: Octokit, octokitOptions: OctokitOptions) {
   events.on(
     "secondary-limit",
     isUsingDeprecatedOnAbuseLimitHandler
-      ? function (...args: [number, ThrottlingOptions, Octokit]) {
+      ? function (...args: [number, OctokitOptions, Octokit]) {
           octokit.log.warn(
             "[@octokit/plugin-throttling] `onAbuseLimit()` is deprecated and will be removed in a future release of `@octokit/plugin-throttling`, please use the `onSecondaryRateLimit` handler instead"
           );
```

File: src/types.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+export type SecondaryLimitHandler = {
+  onSecondaryRateLimit: LimitHandler;
+};
+
+export type ThrottlingOptionsBase = {
+  enabled?: boolean;
+  Bottleneck?: any;
+  id?: string;
+  timeout?: number;
+  connection?: any;
```
Comment (wolfy1339) [inline src/types.ts:32]:
```suggestion
  Bottleneck?: typeof Bottleneck;
  id?: string;
  timeout?: number;
  connection?: Bottleneck.RedisConnection;
```

File: src/route-matcher.ts [hunk 8]
```diff
@@ -1,16 +1,16 @@
-// @ts-ignore
+// @ts-expect-error
 export function routeMatcher(paths) {
```
Comment (wolfy1339) [inline src/route-matcher.ts:2]:
```suggestion
export function routeMatcher(paths: string) {
```

With this simple addition, we can remove all the `// @ts-expect-error` in this file

File: src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts:57]: I just added a new file `test/typescript-validate.ts` (inspired by the one we have in [@octokit/webhooks.js](https://github.com/octokit/webhooks.js/blob/master/test/typescript-validate.ts)) so we can validate every time we modify tests are not introducing a regression.

[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts:57]: I just pushed this branch: `improve-types-with-wolfy-suggestion` with your changes.
Also, when doing cmd+click to `OctokitOptions` there is what I get so I understand I'm updating `OctokitOptions` properly yn my `node_modules`:

image

What I obtain as unexpected TS errors are these ones:
image

@wolfy1339 can you check if those type validations work in your machine?

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts:57]: The type tests work as expected, the same 2 fail for me.

The `// @ts-expect-error` flags should be above the `throttle` key of the options object, in order to silence the error.

[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts:57]: > The type tests work as expected, the same 2 fail for me.
> 
> The `// @ts-expect-error` flags should be above the `throttle` key of the options object, in order to silence the error.

Yes, for the second case in the screenshot yes, you are right. For the first one, in my local, TS does not complain in `improve-types-with-wolfy-suggestion` branch as expected. Does it complain for you?

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts:57]: The `onSecondaryLimit() and onAbuseLimit() should be a TS Error` test does not complain for me.

[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts:57]: Can I see how your modified `node_modules` file looks like?

This is mine:

<details>
  <summary>`node_modules/@octokit/core/dist-types/types.d.ts`</summary>


```javascript
export declare interface OctokitOptions {
    authStrategy?: any;
    auth?: any;
    userAgent?: string;
    previews?: string[];
     baseUrl?: string;
    log?: {
        debug: (message: string) => unknown;
        info: (message: string) => unknown;
        warn: (message: string) => unknown;
        error: (message: string) => unknown;
    };
    request?: OctokitTypes.RequestRequestOptions;
    timeZone?: string;
    [option: string]: any;
}
  ```
</details>

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts:57]:
```ts
export declare interface OctokitOptions {
    authStrategy?: any;
    auth?: any;
    userAgent?: string;
    previews?: string[];
    baseUrl?: string;
    log?: {
        debug: (message: string) => unknown;
        info: (message: string) => unknown;
        warn: (message: string) => unknown;
        error: (message: string) => unknown;
    };
    request?: OctokitTypes.RequestRequestOptions;
    timeZone?: string;
    [option: string]: any;
}```

[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts:57]: Same file (check path)

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts:57]: `node_modules/@octokit/core/dist-types/types.d.ts`

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts:57]: The version of TS that VSCode uses, and the one present in the repo, are 2 different versions.

VSCode: `4.6.2`
Repo: `4.4.3`

[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts:57]: > `node_modules/@octokit/core/dist-types/types.d.ts`

Strange. I guess we will need to sole this issue (https://github.com/octokit/core.js/issues/451), update this branch with the new version, apply your suggestion and push it here so CI tell us if it works.

**Comment by oscard0m** [comment]
Blocked by https://github.com/octokit/core.js/issues/451 resolution to update `@octokit/core` dependency with latest `interface OctokitOptions` change ()

File: src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts]: @wolfy1339 updated this branch with `@octokit/core` and CI is running `validate:ts` now. Seems it is getting the same error I'm having. When you run `npm run validate:ts` in your local, does it work?

In case it doesn't, I guess we will need to rethink the typing (maybe with the approach I was trying to use?)

[the remark this answers is not in the document]
Comment (wolfy1339) [reply src/index.ts]: The `onSecondaryLimit() and onAbuseLimit() should be a TS Error` still doesn't error for me, with clean `node_modules`.
```
$ npm run validate:ts

> @octokit/plugin-throttling@0.0.0-development validate:ts
> tsc --noEmit --noImplicitAny --target es2020 --esModuleInterop --moduleResolution node test/typescript-validate.ts

test/typescript-validate.ts:26:3 - error TS2578: Unused '@ts-expect-error' directive.

26   // @ts-expect-error
     ~~~~~~~~~~~~~~~~~~~


Found 1 error.
```

[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts]: @G-Rath Would you have time to assist us on this?

Right now we are having problems making the types a bit more rigid.

The only way I know to make them more rigid is using method overload, but this makes [TS complain](https://github.com/octokit/plugin-throttling.js/pull/457#discussion_r824273665). Is there a different way to make the types more rigid and at the same time using the Declaration merging you suggested?

[the remark this answers is not in the document]
Comment (G-Rath) [reply src/index.ts]: Try this:

```
export type AbuseLimitHandler = {
  /**
   * @deprecated "[@octokit/plugin-throttling] `onAbuseLimit()` is deprecated and will be removed in a future release of `@octokit/plugin-throttling`, please use the `onSecondaryRateLimit` handler instead"
   */
  onAbuseLimit: LimitHandler;
  onSecondaryRateLimit?: never;
};

export type SecondaryLimitHandler = {
  onAbuseLimit?: never;
  onSecondaryRateLimit: LimitHandler;
};
```

iirc this is because objects in TypeScript are "open", so `x | y` means that the object must satisfy at least the type for `x | y`, not that it can't satisfy the type of both.

You can (sort of*) enforce this by using optional + `never`: since each type now defines the inverse property as having to have type `never`, defining both results in an error because such an object does not satisfy either type in the union.

*: from what I understand this is technically not the same as what we really want, e.g. it's still defining an actual property, so IDEs might still include it in their autocomplete lists, and TypeScripts errors about this will be something like "this property should be type never" rather than a more helpful "you can't implement this property _and_ that property", but it should be close enough to what you want (and it's as close as you're able to get currently anyway)

[the remark this answers is not in the document]
Comment (oscard0m) [reply src/index.ts]: Thanks once again @G-Rath. Since this will be a temporal typing until we deprecate `onAbuse` I think is more than enough.

**Review by wolfy1339 — APPROVED** [review]

**Review by gr2m — APPROVED** [review]
Great PR, thanks for adding type checks!
